pkg/services/ngalert/schedule/schedule_test.go

```diff
@@ -222,6 +228,78 @@ func TestAlertingTicker(t *testing.T) {
 	})
 }
 
+func TestSendingToExternalAlertmanager(t *testing.T) {
+	dbstore := tests.SetupTestEnv(t, 1)
+	t.Cleanup(registry.ClearOverrides)
+
+	// create alert rule with one second interval
+	alertRule := tests.CreateTestAlertRule(t, dbstore, 1)
+
+	fakeAM := newFakeExternalAlertmanager(t)
+	defer fakeAM.Close()
+
+	// First, let's create an admin configuration that holds an alertmanager.
+	adminConfig := &models.AdminConfiguration{OrgID: 1, Alertmanagers: []string{fakeAM.server.URL}}
+	cmd := store.UpdateAdminConfigurationCmd{AdminConfiguration: adminConfig}
+	require.NoError(t, dbstore.UpdateAdminConfiguration(cmd))
+
+	mockedClock := clock.NewMock()
+	baseInterval := time.Second
+
+	logger := log.New("ngalert schedule test")
+	schedCfg := schedule.SchedulerCfg{
+		C:                mockedClock,
+		BaseInterval:     baseInterval,
+		MaxAttempts:      1,
+		Evaluator:        eval.Evaluator{Cfg: &setting.Cfg{ExpressionsEnabled: true}, Log: logger},
+		RuleStore:        dbstore,
+		InstanceStore:    dbstore,
+		AdminConfigStore: dbstore,
+		Notifier:         &fakeNotifier{},
+		Logger:           logger,
+		Metrics:          metrics.NewMetrics(prometheus.NewRegistry()),
+	}
+	st := state.NewManager(schedCfg.Logger, nilMetrics, dbstore, dbstore)
+	sched := schedule.NewScheduler(schedCfg, nil, "http://localhost", st)
+
+	ctx := context.Background()
+
+	// Make sure we sync the configuration at least once before the evaluation happens to guarantee the sender is running
+	// when the first alert triggers.
+	require.NoError(t, sched.SyncAndApplyConfigFromDatabase())
+
+	// Then, ensure we've discovered the Alertmanager.
+	require.Eventually(t, func() bool {
+		return len(sched.AlertmanagersFor(1)) == 1 && len(sched.DroppedAlertmanagersFor(1)) == 0
+	}, 10*time.Second, 200*time.Millisecond)
+
+	go func() {
+		schedule.PollingInterval = 1 * time.Second
+		err := sched.Run(ctx)
+		require.NoError(t, err)
+	}()
+
+	// With everything up and running, let's advance the time to make sure we get at least one alert iteration.
+	mockedClock.Add(2 * time.Second)
+
+	// Eventually, our Alertmanager should have received at least one alert.
+	require.Eventually(t, func() bool {
```

Comment:
Do we still need `require.Eventually()` since we have used a mockedClock?

Reply:
After the alert being sent to the Prometheus notifier, it goes into a queue, given this is all run async by the time we check the alert would not have been delivered yet which would result in a flaky test given this is all async.

Reply:
thanks for clarifying it, GTK!
my next question is: 10 seconds is enough time to be sure that we don't end up with a flaky test or should we increase it a bit?

Reply:
The total runtime (in my machine which at the moment is quite slow) is ~5s - I gave it 10s to ensure CI has more than enough leeway. Does that seem right to you?